src/lib/check.ml

```diff
@@ -0,0 +1,157 @@
+module D = Domain
+module Syn = Syntax
+type env_entry =
+    Term of {term : D.t; tp : D.t}
+  | TopLevel of {term : D.t; tp : D.t}
+type env = env_entry list
+
+let add_term ~term ~tp env = Term {term; tp} :: env
+
+type error =
+    Cannot_synth_term of Syn.t
+  | Type_mismatch of D.t * D.t
+  | Expecting_universe of D.t
+  | Misc of string
+
+let pp_error = function
+  | Cannot_synth_term t -> "Cannot synthesize the type of:\n" ^ Syn.pp t
+  | Type_mismatch (t1, t2) -> "Cannot equate\n" ^ D.pp t1 ^ " with\n" ^ D.pp t2
+  | Expecting_universe d -> "Expected some universe but found\n" ^ D.pp d
+  | Misc s -> s
+
+exception Type_error of error
+
+let tp_error e = raise (Type_error e)
+
+let env_to_sem_env =
+  List.map
+    (function
+      | TopLevel {term; _} -> term
+      | Term {term; _} -> term)
+
+module S = Set.Make(struct type t = int;; let compare = compare end)
+
+let get_var env n = match List.nth env n with
+  | Term {term = _; tp} -> tp
+  | TopLevel {tp; _} -> tp
+
+let assert_subtype size t1 t2 =
+  if Nbe.check_subtype size t1 t2
+  then ()
+  else tp_error (Type_mismatch (t1, t2))
+
+let rec check ~env ~size ~term ~tp =
```

Comment:
I should consider some of these modes more carefully. For instance, I can definitely synth `Nat`. No real reason not to.